File: frontend/lib/anchoredPatch.ts
import { matchExperienceCompany, matchProjectName } from "@/lib/applyResumePatch";
import type { BlockingIssue, IssueAnchor, ResumePatch, TailoredResumeOutput } from "@/lib/api";

export function resolveBulletAtAnchor(
  resume: TailoredResumeOutput,
  anchor: IssueAnchor,
): {
  section: IssueAnchor["section"];
  company?: string;
  project_name?: string;
  institution?: string;
  bullet_old?: string;
  project_bullet_old?: string;
  education_bullet_old?: string;
} | null {
  const bulletIndex = anchor.bullet_index;
  if (bulletIndex == null) return null;

  if (anchor.section === "experience") {
    const entry = resume.experience[anchor.entry_index];
    const bullet = entry?.bullets[bulletIndex];
    const company = entry?.company?.trim();
    if (!entry || !company || !bullet) return null;
    return { section: "experience", company, bullet_old: bullet };
  }

  if (anchor.section === "projects") {
    const entry = resume.projects[anchor.entry_index] as { name?: string; bullets?: string[] } | undefined;
    const bullet = entry?.bullets?.[bulletIndex];
    const name = entry?.name?.trim();
    if (!entry || !name || !bullet) return null;
    return { section: "projects", project_name: name, project_bullet_old: bullet };
  }

  const entry = resume.education[anchor.entry_index];
  const bullet = entry?.bullets?.[bulletIndex];
  const institution = entry?.institution?.trim();
  if (!entry || !institution || !bullet) return null;
  return { section: "education", institution, education_bullet_old: bullet };
}

export function hydratePatchFromAnchor(
  resume: TailoredResumeOutput,
  patch: ResumePatch,
  anchor?: IssueAnchor | null,
): ResumePatch {
  if (!anchor) return patch;
  const resolved = resolveBulletAtAnchor(resume, anchor);
  if (!resolved) {
    return { ...patch, anchor };
  }

  if (resolved.section === "experience") {
    return {
      ...patch,
      anchor,
      section: "experience",
      company: resolved.company,
      bullet_old: resolved.bullet_old,
    };
  }
  if (resolved.section === "projects") {
    return {
      ...patch,
      anchor,
      section: "projects",
      project_name: resolved.project_name,
      project_bullet_old: resolved.project_bullet_old,
    };
  }
  return {
    ...patch,
    anchor,
    section: "education",
    institution: resolved.institution,
    education_bullet_old: resolved.education_bullet_old,
  };
}

function normalizeBulletText(value: string): string {
  return value.trim().replace(/\s+/g, " ");
}

function bulletHintMatches(candidate: string, hint: string): boolean {
  const bullet = normalizeBulletText(candidate);
  const needle = normalizeBulletText(hint.replace(/…$/, ""));
  if (!bullet || !needle) return false;
  if (bullet === needle) return true;
  if (bullet.startsWith(needle) || needle.startsWith(bullet)) return true;
  const prefixLen = 16;
  if (needle.length >= prefixLen && bullet.includes(needle.slice(0, prefixLen))) return true;
  return false;
}

function bulletHintFromIssue(issue: BlockingIssue): string | null {
  const text = issue.suggestion.trim();
  if (!text) return null;
  const hint = text.split(":").pop()?.trim().replace(/…$/, "").trim() ?? "";
  return hint.length >= 8 ? hint : null;
}

export function inferAnchorFromIssue(
  resume: TailoredResumeOutput,
  issue: BlockingIssue,
): IssueAnchor | null {
  if (issue.anchor) return issue.anchor;
  const hint = bulletHintFromIssue(issue);
  if (!hint) return null;

  const matches: IssueAnchor[] = [];
  resume.experience.forEach((entry, entryIndex) => {
    entry.bullets.forEach((bullet, bulletIndex) => {
      if (bulletHintMatches(bullet, hint)) {
        matches.push({ section: "experience", entry_index: entryIndex, bullet_index: bulletIndex });
      }
    });
  });
  (resume.projects ?? []).forEach((entry, entryIndex) => {
    const bullets = Array.isArray((entry as { bullets?: string[] }).bullets)
      ? (entry as { bullets: string[] }).bullets
      : [];
    bullets.forEach((bullet, bulletIndex) => {
      if (bulletHintMatches(bullet, hint)) {
        matches.push({ section: "projects", entry_index: entryIndex, bullet_index: bulletIndex });
      }
    });
  });
  resume.education.forEach((entry, entryIndex) => {
    (entry.bullets ?? []).forEach((bullet, bulletIndex) => {
      if (bulletHintMatches(bullet, hint)) {
        matches.push({ section: "education", entry_index: entryIndex, bullet_index: bulletIndex });
      }
    });
  });
  return matches.length === 1 ? matches[0]! : null;
}

export function enrichIssuesWithInferredAnchors(
  resume: TailoredResumeOutput,
  issues: BlockingIssue[],
): BlockingIssue[] {
  return issues.map((issue) => {
    const anchor = inferAnchorFromIssue(resume, issue);
    if (!anchor || issue.anchor) return issue;
    return { ...issue, anchor };
  });
}

function anchorInIssues(anchor: IssueAnchor, issues: BlockingIssue[]): boolean {
  return issues.some(
    (issue) =>
      issue.anchor?.section === anchor.section &&
      issue.anchor.entry_index === anchor.entry_index &&
      issue.anchor.bullet_index === anchor.bullet_index,
  );
}

function patchConflictsWithAnchor(
  resume: TailoredResumeOutput,
  patch: ResumePatch,
  anchor: IssueAnchor,
): boolean {
  const resolved = resolveBulletAtAnchor(resume, anchor);
  if (!resolved) return true;
  if (resolved.section === "experience" && patch.section === "experience") {
    const company = patch.company?.trim();
    return Boolean(company && !matchExperienceCompany(company, resolved.company ?? ""));
  }
  if (resolved.section === "projects" && patch.section === "projects") {
    const projectName = patch.project_name?.trim();
    return Boolean(
      projectName && !matchProjectName(projectName, resolved.project_name ?? ""),
    );
  }
  if (resolved.section === "education" && patch.section === "education") {
    const institution = patch.institution?.trim();
    return Boolean(
      institution &&
        institution.toLowerCase() !== (resolved.institution ?? "").toLowerCase(),
    );
  }
  return false;
}

function patchContentMatchesAnchor(
  resume: TailoredResumeOutput,
  patch: ResumePatch,
  anchor: IssueAnchor,
): boolean {
  const resolved = resolveBulletAtAnchor(resume, anchor);
  if (!resolved) return false;
  if (resolved.section === "experience" && patch.section === "experience") {
    const company = patch.company?.trim();
    const bulletOld = patch.bullet_old?.trim();
    if (company && !matchExperienceCompany(company, resolved.company ?? "")) return false;
    if (bulletOld && resolved.bullet_old && !bulletHintMatches(resolved.bullet_old, bulletOld)) {
      return false;
    }
    return true;
  }
  if (resolved.section === "projects" && patch.section === "projects") {
    const projectName = patch.project_name?.trim();
    const bulletOld = patch.project_bullet_old?.trim();
    if (projectName && !matchProjectName(projectName, resolved.project_name ?? "")) return false;
    if (
      bulletOld &&
      resolved.project_bullet_old &&
      !bulletHintMatches(resolved.project_bullet_old, bulletOld)
    ) {
      return false;
    }
    return true;
  }
  if (resolved.section === "education" && patch.section === "education") {
    const institution = patch.institution?.trim();
    const bulletOld = patch.education_bullet_old?.trim();
    if (institution && institution.toLowerCase() !== (resolved.institution ?? "").toLowerCase()) {
      return false;
    }
    if (
      bulletOld &&
      resolved.education_bullet_old &&
      !bulletHintMatches(resolved.education_bullet_old, bulletOld)
    ) {
      return false;
    }
    return true;
  }
  return false;
}

export function matchAnchorForPatch(
  resume: TailoredResumeOutput,
  patch: ResumePatch,
  issues: BlockingIssue[],
): IssueAnchor | null {
  const matches: IssueAnchor[] = [];
  for (const issue of issues) {
    if (!issue.anchor) continue;
    const resolved = resolveBulletAtAnchor(resume, issue.anchor);
    if (!resolved) continue;
    if (resolved.section === "experience" && patch.section === "experience") {
      const company = patch.company?.trim();
      if (company && matchExperienceCompany(company, resolved.company ?? "")) {
        matches.push(issue.anchor);
        continue;
      }
      if (patch.bullet_new && !company) {
        matches.push(issue.anchor);
      }
    } else if (resolved.section === "projects" && patch.section === "projects") {
      const projectName = patch.project_name?.trim();
      if (projectName && matchProjectName(projectName, resolved.project_name ?? "")) {
        matches.push(issue.anchor);
        continue;
      }
      if (patch.project_bullet_new && !projectName) {
        matches.push(issue.anchor);
      }
    }
  }
  return matches.length === 1 ? matches[0] : null;
}

function anchorForPatchIndex(
  resume: TailoredResumeOutput,
  patch: ResumePatch,
  issues: BlockingIssue[],
  patchIndex: number,
): IssueAnchor | null {
  const matched = matchAnchorForPatch(resume, patch, issues);
  if (matched) return matched;
  const positional = issues[patchIndex]?.anchor;
  if (positional) {
    if (issues.length === 1 || !patchConflictsWithAnchor(resume, patch, positional)) {
      return positional;
    }
  }
  return patch.anchor && anchorInIssues(patch.anchor, issues) ? patch.anchor : null;
}

export function hydratePatchesFromIssues(
  resume: TailoredResumeOutput,
  patches: ResumePatch[],
  issues: BlockingIssue[],
): ResumePatch[] {
  const enriched = enrichIssuesWithInferredAnchors(resume, issues);
  if (!enriched.some((issue) => issue.anchor)) return patches;
  return patches.map((patch, index) =>
    hydratePatchFromAnchor(
      resume,
      patch,
      anchorForPatchIndex(resume, patch, enriched, index),
    ),
  );
}

export function sourceIssueKeysForPatches(
  resume: TailoredResumeOutput,
  patches: ResumePatch[],
  issues: BlockingIssue[],
): (string | undefined)[] {
  const enriched = enrichIssuesWithInferredAnchors(resume, issues);
  return patches.map((patch) => {
    const anchor = patch.anchor ?? matchAnchorForPatch(resume, patch, enriched);
    if (!anchor) return undefined;
    const issue = enriched.find(
      (candidate) =>
        candidate.anchor?.section === anchor.section &&
        candidate.anchor.entry_index === anchor.entry_index &&
        candidate.anchor.bullet_index === anchor.bullet_index,
    );
    return issue ? issueAnchorKey(issue) : undefined;
  });
}

export function buildBatchChatMessage(
  issues: BlockingIssue[],
  tailored: TailoredResumeOutput | null,
): string {
  const lines = issues.map((issue, index) => {
    const resolved =
      issue.anchor && tailored ? resolveBulletAtAnchor(tailored, issue.anchor) : null;
    const exactBullet =
      resolved?.bullet_old ??
      resolved?.project_bullet_old ??
      resolved?.education_bullet_old;
    const location =
      resolved?.company != null
        ? `company="${resolved.company}"`
        : resolved?.project_name != null
          ? `project="${resolved.project_name}"`
          : resolved?.institution != null
            ? `institution="${resolved.institution}"`
            : null;
    const bulletLine = exactBullet
      ? `\n   EXACT current bullet: "${exactBullet}"`
      : "";
    const locationLine = location ? `\n   Location: ${location}` : "";
    return `${index + 1}. [${issue.category}] ${issue.description}\n   Suggestion: ${issue.suggestion}${locationLine}${bulletLine}`;
  });
  return `Address these ${issues.length} issues in my resume in a single round of edits:\n\n${lines.join("\n\n")}\n\nReturn one patch per issue above, in order. Use the EXACT current bullet text for bullet_old when provided.`;
}

export function issueAnchorKey(issue: BlockingIssue): string {
  if (!issue.anchor) {
    return `${issue.category}|${issue.description}|${issue.suggestion}`;
  }
  const { section, entry_index, bullet_index } = issue.anchor;
  return `anchor:${section}:${entry_index}:${bullet_index ?? "x"}`;
}

File: backend/app/services/anchored_patch.py
"""Resolve ATS issue anchors to exact resume bullets for chat patch hydration."""

from __future__ import annotations

from typing import Any

from app.models.chat import ResumePatch
from app.models.qa import BlockingIssue, IssueAnchor


def _normalize_name(value: str) -> str:
    return " ".join(value.strip().lower().split())


def _names_match(a: str, b: str) -> bool:
    left = _normalize_name(a)
    right = _normalize_name(b)
    if not left or not right:
        return False
    return left == right or left in right or right in left


def _normalize_bullet(value: str) -> str:
    return " ".join(value.strip().split())


# Prefix match only when the hint is long enough to avoid cross-bullet collisions.
_ANCHOR_HINT_PREFIX_LEN = 16


def _bullet_hint_matches(candidate: str, hint: str) -> bool:
    bullet = _normalize_bullet(candidate)
    needle = _normalize_bullet(hint.rstrip("…"))
    if not bullet or not needle:
        return False
    if bullet == needle:
        return True
    if bullet.startswith(needle) or needle.startswith(bullet):
        return True
    if (
        len(needle) >= _ANCHOR_HINT_PREFIX_LEN
        and needle[:_ANCHOR_HINT_PREFIX_LEN] in bullet
    ):
        return True
    return False


def _bullet_hint_from_issue(issue: BlockingIssue) -> str | None:
    text = issue.suggestion.strip()
    if not text:
        return None
    hint = text.rsplit(":", 1)[-1].strip().rstrip("…").strip()
    return hint if len(hint) >= 8 else None


def infer_anchor_from_issue(
    resume: dict[str, Any],
    issue: BlockingIssue,
) -> IssueAnchor | None:
    """Best-effort anchor when Phase 4 stored suggestion text but no anchor."""
    if issue.anchor is not None:
        return issue.anchor
    hint = _bullet_hint_from_issue(issue)
    if not hint:
        return None

    matches: list[IssueAnchor] = []
    for entry_index, entry in enumerate(resume.get("experience") or []):
        for bullet_index, bullet in enumerate(entry.get("bullets") or []):
            if _bullet_hint_matches(str(bullet), hint):
                matches.append(
                    IssueAnchor(
                        section="experience",
                        entry_index=entry_index,
                        bullet_index=bullet_index,
                    )
                )
    for entry_index, entry in enumerate(resume.get("projects") or []):
        for bullet_index, bullet in enumerate(entry.get("bullets") or []):
            if _bullet_hint_matches(str(bullet), hint):
                matches.append(
                    IssueAnchor(
                        section="projects",
                        entry_index=entry_index,
                        bullet_index=bullet_index,
                    )
                )
    for entry_index, entry in enumerate(resume.get("education") or []):
        for bullet_index, bullet in enumerate(entry.get("bullets") or []):
            if _bullet_hint_matches(str(bullet), hint):
                matches.append(
                    IssueAnchor(
                        section="education",
                        entry_index=entry_index,
                        bullet_index=bullet_index,
                    )
                )
    return matches[0] if len(matches) == 1 else None


def enrich_issues_with_inferred_anchors(
    resume: dict[str, Any],
    issues: list[BlockingIssue],
) -> list[BlockingIssue]:
    enriched: list[BlockingIssue] = []
    for issue in issues:
        anchor = infer_anchor_from_issue(resume, issue)
        if anchor is None or issue.anchor is not None:
            enriched.append(issue)
            continue
        enriched.append(issue.model_copy(update={"anchor": anchor}))
    return enriched


def _anchor_in_issues(anchor: IssueAnchor, issues: list[BlockingIssue]) -> bool:
    for issue in issues:
        if issue.anchor is None:
            continue
        if (
            issue.anchor.section == anchor.section
            and issue.anchor.entry_index == anchor.entry_index
            and issue.anchor.bullet_index == anchor.bullet_index
        ):
            return True
    return False


def resolve_bullet_at_anchor(
    resume: dict[str, Any],
    anchor: IssueAnchor,
) -> dict[str, str] | None:
    """Return canonical section fields for a bullet at anchor, or None if stale."""
    bullet_index = anchor.bullet_index
    if bullet_index is None:
        return None

    if anchor.section == "experience":
        entries = resume.get("experience") or []
        if anchor.entry_index >= len(entries):
            return None
        entry = entries[anchor.entry_index]
        bullets = entry.get("bullets") or []
        if bullet_index >= len(bullets):
            return None
        company = str(entry.get("company") or "").strip()
        if not company:
            return None
        return {
            "section": "experience",
            "company": company,
            "bullet_old": str(bullets[bullet_index]),
        }

    if anchor.section == "projects":
        entries = resume.get("projects") or []
        if anchor.entry_index >= len(entries):
            return None
        entry = entries[anchor.entry_index]
        bullets = entry.get("bullets") or []
        if bullet_index >= len(bullets):
            return None
        name = str(entry.get("name") or "").strip()
        if not name:
            return None
        return {
            "section": "projects",
            "project_name": name,
            "project_bullet_old": str(bullets[bullet_index]),
        }

    if anchor.section == "education":
        entries = resume.get("education") or []
        if anchor.entry_index >= len(entries):
            return None
        entry = entries[anchor.entry_index]
        bullets = entry.get("bullets") or []
        if bullet_index >= len(bullets):
            return None
        institution = str(entry.get("institution") or "").strip()
        if not institution:
            return None
        return {
            "section": "education",
            "institution": institution,
            "education_bullet_old": str(bullets[bullet_index]),
        }

    return None


def build_anchor_target_block(
    resume: dict[str, Any],
    issues: list[BlockingIssue],
) -> str:
    """Human + machine-readable targets for the chat agent."""
    lines = [
        "ANCHORED FIX TARGETS (rewrite ONLY bullet_new — bullet_old is filled server-side):",
    ]
    for i, issue in enumerate(issues, start=1):
        if not issue.anchor:
            continue
        resolved = resolve_bullet_at_anchor(resume, issue.anchor)
        if not resolved:
            lines.append(
                f"{i}. [{issue.category}] {issue.description} — anchor stale, skip patch"
            )
            continue
        if resolved["section"] == "experience":
            lines.append(
                f'{i}. section=experience company="{resolved["company"]}"\n'
                f'   EXACT current bullet: "{resolved["bullet_old"]}"\n'
                f"   Fix intent: {issue.suggestion}"
            )
        elif resolved["section"] == "projects":
            lines.append(
                f'{i}. section=projects project_name="{resolved["project_name"]}"\n'
                f'   EXACT current bullet: "{resolved["project_bullet_old"]}"\n'
                f"   Fix intent: {issue.suggestion}"
            )
        else:
            lines.append(
                f'{i}. section=education institution="{resolved["institution"]}"\n'
                f'   EXACT current bullet: "{resolved["education_bullet_old"]}"\n'
                f"   Fix intent: {issue.suggestion}"
            )
    return "\n".join(lines)


def hydrate_patch_from_anchor(
    resume: dict[str, Any],
    patch: ResumePatch,
    anchor: IssueAnchor | None,
) -> ResumePatch:
    """Stamp bullet_old / company / project_name from live resume at anchor."""
    if anchor is None:
        return patch
    resolved = resolve_bullet_at_anchor(resume, anchor)
    if not resolved:
        patch.anchor = anchor
        return patch

    patch.anchor = anchor
    if resolved["section"] == "experience":
        patch.section = "experience"
        patch.company = resolved["company"]
        patch.bullet_old = resolved["bullet_old"]
    elif resolved["section"] == "projects":
        patch.section = "projects"
        patch.project_name = resolved["project_name"]
        patch.project_bullet_old = resolved["project_bullet_old"]
    elif resolved["section"] == "education":
        patch.section = "education"
        patch.institution = resolved["institution"]
        patch.education_bullet_old = resolved["education_bullet_old"]
    return patch


def _patch_conflicts_with_anchor(
    resume: dict[str, Any],
    patch: ResumePatch,
    anchor: IssueAnchor,
) -> bool:
    """True when patch names a different company/project than the anchor location."""
    resolved = resolve_bullet_at_anchor(resume, anchor)
    if not resolved:
        return True
    if resolved["section"] == "experience" and patch.section == "experience":
        company = (patch.company or "").strip()
        return bool(company and not _names_match(company, resolved["company"]))
    if resolved["section"] == "projects" and patch.section == "projects":
        project_name = (patch.project_name or "").strip()
        return bool(
            project_name and not _names_match(project_name, resolved["project_name"])
        )
    if resolved["section"] == "education" and patch.section == "education":
        institution = (patch.institution or "").strip()
        return bool(
            institution and not _names_match(institution, resolved["institution"])
        )
    return False


def _patch_content_matches_anchor(
    resume: dict[str, Any],
    patch: ResumePatch,
    anchor: IssueAnchor,
) -> bool:
    """True when patch company/name/bullet_old align with the anchor location."""
    resolved = resolve_bullet_at_anchor(resume, anchor)
    if not resolved:
        return False
    if resolved["section"] == "experience" and patch.section == "experience":
        company = (patch.company or "").strip()
        bullet_old = (patch.bullet_old or "").strip()
        if company and not _names_match(company, resolved["company"]):
            return False
        if bullet_old and not _bullet_hint_matches(resolved["bullet_old"], bullet_old):
            return False
        return True
    if resolved["section"] == "projects" and patch.section == "projects":
        project_name = (patch.project_name or "").strip()
        bullet_old = (patch.project_bullet_old or "").strip()
        if project_name and not _names_match(project_name, resolved["project_name"]):
            return False
        if bullet_old and not _bullet_hint_matches(resolved["project_bullet_old"], bullet_old):
            return False
        return True
    if resolved["section"] == "education" and patch.section == "education":
        institution = (patch.institution or "").strip()
        bullet_old = (patch.education_bullet_old or "").strip()
        if institution and not _names_match(institution, resolved["institution"]):
            return False
        if bullet_old and not _bullet_hint_matches(resolved["education_bullet_old"], bullet_old):
            return False
        return True
    return False


def match_anchor_for_patch(
    resume: dict[str, Any],
    patch: ResumePatch,
    target_issues: list[BlockingIssue],
) -> IssueAnchor | None:
    """Match a patch to the anchored issue it targets (not positional)."""
    matches: list[IssueAnchor] = []
    for issue in target_issues:
        if issue.anchor is None:
            continue
        resolved = resolve_bullet_at_anchor(resume, issue.anchor)
        if not resolved:
            continue
        if resolved["section"] == "experience" and patch.section == "experience":
            company = (patch.company or "").strip()
            if company and _names_match(company, resolved["company"]):
                matches.append(issue.anchor)
                continue
            if patch.bullet_new and not company:
                matches.append(issue.anchor)
        elif resolved["section"] == "projects" and patch.section == "projects":
            project_name = (patch.project_name or "").strip()
            if project_name and _names_match(project_name, resolved["project_name"]):
                matches.append(issue.anchor)
                continue
            if patch.project_bullet_new and not project_name:
                matches.append(issue.anchor)
    if len(matches) == 1:
        return matches[0]
    return None


def _anchor_for_patch_index(
    resume: dict[str, Any],
    patch: ResumePatch,
    target_issues: list[BlockingIssue],
    patch_index: int,
) -> IssueAnchor | None:
    """Bind a chat patch to its ATS anchor — content match first, guarded positional."""
    anchor = match_anchor_for_patch(resume, patch, target_issues)
    if anchor is not None:
        return anchor

    if patch_index < len(target_issues):
        positional = target_issues[patch_index].anchor
        if positional is not None:
            if len(target_issues) == 1:
                return positional
            if not _patch_conflicts_with_anchor(resume, patch, positional):
                return positional

    if patch.anchor is not None and _anchor_in_issues(patch.anchor, target_issues):
        return patch.anchor
    return None


def hydrate_chat_patches(
    resume: dict[str, Any],
    patches: list[ResumePatch],
    target_issues: list[BlockingIssue],
) -> list[ResumePatch]:
    """Attach anchors and canonical bullet_old to LLM patches."""
    issues = enrich_issues_with_inferred_anchors(resume, target_issues)
    if not any(issue.anchor is not None for issue in issues):
        return patches

    hydrated: list[ResumePatch] = []
    for index, patch in enumerate(patches):
        anchor = _anchor_for_patch_index(resume, patch, issues, index)
        hydrated.append(hydrate_patch_from_anchor(resume, patch, anchor))

    return hydrated
